### PR TITLE
CUMULUS-2568 -- Fix reconciliation tests 

### PR DESCRIPTION
**Pull request by Jkovarik**
**Summary:** Summary of changes

This PR was originally opened to address a cleanup failure where the 'static' collection (MYD13Q1_006) being used for the reconciliation report test wasn't cleaning itself up and then halting a stack from running the test without manual intervention. 

Investigating that issue also showed other ancillary cleanup was failing silently/consistently.    

These changes attempt to address that and related failures introduced via https://github.com/nasa/cumulus/pull/2302

Addresses [CUMULUS-2568](https://bugs.earthdata.nasa.gov/browse/CUMULUS-2568)

## Changes
  - Add `deletePdr`/PDR deletion functionality to `@cumulus/api-client/pdrs`
  - Add `removeCollectionAndAllDependencies` to integration test helpers
  - Update reconciliation report integration test to have better cleanup/failure
    behavior

## PR Checklist

- [x] Update CHANGELOG
- [x] Unit tests
- [x] Ad-hoc testing - Deploy changes and test manually
- [x] Integration tests

### CODE REVIEW

File: example/spec/helpers/Collections.js [hunk 1]
```diff
@@ -0,0 +1,74 @@
+const { listGranules, removePublishedGranule, deleteGranule } = require('@cumulus/api-client/granules');
+const { getPdrs, deletePdr } = require('@cumulus/api-client/pdrs');
+const { deleteExecution, getExecutions } = require('@cumulus/api-client/executions');
+const { deleteCollection } = require('@cumulus/api-client/collections');
+const { constructCollectionId } = require('@cumulus/message/Collections');
+
+const removeCollectionAndAllDependencies = async (params) => {
```
Comment (Jkovarik) [inline example/spec/helpers/Collections.js:7]: Helper method to proactively remove all sub-dependent records from a collection.

Needs docstring.

File: packages/api-client/src/pdrs.ts [hunk 2]
```diff
@@ -2,6 +2,34 @@ import { ApiPdr } from '@cumulus/types/api/pdrs';
...
+ *
+ * @param {Object} params            - params
+ * @param {string} params.prefix     - the prefix configured for the stack
+ * @param {string} params.pdrName    - a PDR name
+ * @param {Function} params.callback - async function to invoke the api lambda
+ *                                     that takes a prefix / user payload.  Defaults
+ *                                     to cumulusApiClient.invokeApi
+ * @returns {Promise<Object>}        - the execution fetched by the API
+ */
+export const deletePdr = async (params: {
```
Comment (Jkovarik) [inline packages/api-client/src/pdrs.ts:16]: This is needed to clean up a collection's dependencies, and.... we're going to need it for users eventually.

File: packages/api-client/src/CumulusApiClientError.ts [hunk 3]
```diff
@@ -1,7 +1,12 @@
 export class CumulusApiClientError extends Error {
-  constructor(message: string) {
+  public statusCode: number | undefined;
+  public apiMessage: string | undefined;
+
+  constructor(message: string, statusCode: number, apiMessage: string | undefined) {
```
Comment (Jkovarik) [inline packages/api-client/src/CumulusApiClientError.ts:5]: This update and the update to the invoke method allows for programmatic inspection of the error without needing to parse the usertext message.    This is helpful if you're making a call to invoke the API but want to take different action depending on the api failure *without* modifying the throw/catch behavior. 

I'm open to thinking about explicitly passing in a *list* of status codes and parsing a response, but I think generally the try/catch style is better/more explicit than a bunch of if branch logic for a successful return.

File: example/spec/parallel/createReconciliationReport/CreateReconciliationReportSpec.js [hunk 4]
```diff
@@ -746,36 +765,38 @@ describe('When there are granule differences and granule reconciliation is run',
...
+        responseError = error;
+      }
+      expect(responseError.statusCode).toBe(404);
+      expect(JSON.parse(responseError.apiMessage).message).toBe(`No record found for ${reportRecord.name}`);
     });
   });
 
   afterAll(async () => {
-    console.log(`update granule files back ${publishedGranuleId}`);
+    const activeCollectionId = constructCollectionId(extraCumulusCollection.name, extraCumulusCollection.version);
```
Comment (Jkovarik) [inline example/spec/parallel/createReconciliationReport/CreateReconciliationReportSpec.js:783]: A good portion of this cleanup wasn't working as intended, particularly the extra collection cleanup.   

Given the nature of that cleanup, and this test implementing an API-driven top-down collection deletion seems more maintainable/DRY/etc than the prior approach, however feedback is welcome.

File: example/spec/parallel/createReconciliationReport/CreateReconciliationReportSpec.js [hunk 5]
```diff
@@ -358,12 +339,10 @@ describe('When there are granule differences and granule reconciliation is run',
         publishedGranuleId,
         dbGranuleId,
         cmrGranule,
-        [extraCumulusCollection, extraCumulusCollectionCleanup],
```
Comment (Jkovarik) [inline example/spec/parallel/createReconciliationReport/CreateReconciliationReportSpec.js:361]: See comment in the test cleanup for why this was removed/refactored.

File: example/spec/parallel/createReconciliationReport/CreateReconciliationReportSpec.js [hunk 6]
```diff
@@ -92,12 +89,13 @@ async function setupCollectionAndTestData(config, testSuffix, testDataFolder) {
...
 const ingestAndPublishGranuleExecutionArns = [];
 
+// TODO: [CUMULUS-2567] These should be in a helper, possibly unit tested.
 /**
  * Creates a new test collection with associated granule for testing.
  *
  * @param {string} prefix - stack Prefix
  * @param {string} sourceBucket - testing source bucket
- * @returns {Promise<Array>} A new collection with associated granule and a cleanup function to call after you are finished.
+ * @returns {Promise<Object>}  The collection created
```
Comment (Jkovarik) [inline example/spec/parallel/createReconciliationReport/CreateReconciliationReportSpec.js:98]: Update this docstring

File: example/spec/parallel/createReconciliationReport/CreateReconciliationReportSpec.js [hunk 7]
```diff
@@ -81,6 +77,7 @@ async function setupCollectionAndTestData(config, testSuffix, testDataFolder) {
     '@cumulus/test-data/granules/BROWSE.MYD13Q1.A2002185.h00v09.006.2015149071135.1.jpg',
   ];
 
+  await removeCollectionAndAllDependencies({ config, collection });
```
Comment (Jkovarik) [inline example/spec/parallel/createReconciliationReport/CreateReconciliationReportSpec.js:80]: One of the more vexing failure situations for this test occurs when the static collection used for CMR export fails cleanup - the tests then were failing continually *and* failing on cleanup resulting in a required stack manual intervention for this test to ever pass. 

Short of randomizing this collection, making the cleanup more robust helps, but up-front ensuring the stack state with respect to the MYD13Q1 collection ensures that if bamboo/other factors result in a bad cleanup state the test will start with a relatively fresh slate.    99% of the time this should functionally do nothing.

File: example/spec/helpers/Collections.js (hunk 1 above)
Comment (Jkovarik) [reply example/spec/helpers/Collections.js:19]: That's better: 4f7db6588

File: example/spec/parallel/createReconciliationReport/CreateReconciliationReportSpec.js [hunk 8]
```diff
@@ -553,13 +548,19 @@ describe('When there are granule differences and granule reconciliation is run',
...
-      expect(response.statusCode).toBe(404);
-      expect(JSON.parse(response.body).message).toBe(`No record found for ${reportRecord.name}`);
+      try {
+        await reconciliationReportsApi.getReconciliationReport({
+          prefix: config.stackName,
+          name: reportRecord.name,
+        });
+      } catch (error) {
+        responseError = error;
+      }
```
Comment (markdboyd) [inline example/spec/parallel/createReconciliationReport/CreateReconciliationReportSpec.js:560]: These try/catch will be unnecessary with the changes from #2312

File: example/spec/helpers/Collections.js [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+*            - PDRs
+*            - Executions
+*            - the specified Collection
+* @param {Object} params     - params
+* @param {string} prefix  - Config object containing stackName
+* @param {Object} collection - Cumulus API collection object to delete
+* @return {Promise<undefined>}
+*/
+const removeCollectionAndAllDependencies = async (params) => {
+  const { prefix, collection } = params;
```
Comment (markdboyd) [inline example/spec/helpers/Collections.js:20]: I admit that I'm not a huge fan of complex helper functions like this that do a lot of things vs integration tests making sure they are precise in their cleanup. But I could probably be convinced

File: example/spec/parallel/createReconciliationReport/CreateReconciliationReportSpec.js [hunk 10]
```diff
@@ -397,12 +376,12 @@ describe('When there are granule differences and granule reconciliation is run',
...
     } catch (error) {
       console.log(error);
-      beforeAllFailed = true;
-      throw error;
+      beforeAllFailed = error;
     }
   });
 
   it('prepares the test suite successfully', () => {
+    if (beforeAllFailed) fail(beforeAllFailed);
```
Comment (markdboyd) [inline example/spec/parallel/createReconciliationReport/CreateReconciliationReportSpec.js:384]: Line below is duplicate/unnecessary

**Review by markdboyd — APPROVED** [review]

File: example/spec/parallel/createReconciliationReport/CreateReconciliationReportSpec.js (hunk 10 above)
Comment (Jkovarik) [reply example/spec/parallel/createReconciliationReport/CreateReconciliationReportSpec.js:384]: 2ea7ec979

**Comment by Jkovarik** [comment]
Merging as updated integration test is passing manually.

**Pull request merged**